src/Microsoft.Diagnostics.Monitoring.Egress/FileSystem/FileSystemEgressProvider.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Extensions.Logging;
+using System;
+using System.ComponentModel.DataAnnotations;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.Egress.FileSystem
+{
+    internal class FileSystemEgressProvider :
+        EgressProvider<FileSystemEgressProviderOptions, FileSystemEgressStreamOptions>
+    {
+        public FileSystemEgressProvider(FileSystemEgressProviderOptions options, ILogger logger = null)
+            : base(options, logger)
+        {
+        }
+
+        public override async Task<string> EgressAsync(
+            Func<Stream, CancellationToken, Task> action,
+            string name,
+            FileSystemEgressStreamOptions streamOptions,
+            CancellationToken token)
+        {
+            LogAndValidateOptions(name);
+
+            Logger?.LogDebug("Check if directory exists.");
+            if (!Directory.Exists(Options.DirectoryPath))
+            {
+                Logger?.LogDebug("Start creating directory.");
+                Directory.CreateDirectory(Options.DirectoryPath);
```

Comment:
>CreateDirectory [](start = 26, length = 15)

Do we handle directory creation failures gracefully here?

Reply:
This will 500 the request. I'm wondering if we need to establish a exception contract where the monitoring components need to handle their exceptions and propagate a well-known exception that is handled by requesting layers (e.g. the HTTP API) so that it can differentiate failures due to invalid configuration, invalid user input, communication failures, etc vs failures in the monitoring code itself.

Reply:
The egress providers now wrap all operational exceptions (exceptions not caused by programming error) with a new exception type that is handled at the HTTP API layer.